ovirt_imageio/client/_upload.py

```diff
@@ -0,0 +1,169 @@
+# ovirt-imageio
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+"""
+Upload commands.
+"""
+
+import os
+from contextlib import closing
+from collections import namedtuple
+
+from . import _api
+from . import _options
+from . import _ovirt
+from . import _ui
+
+
+class DiskInfo(namedtuple(
+        "DiskInfo", "name,initial_size,provisioned_size,content_type,format")):
+    """
+    Container class for attributes that the uploaded disk will have.
+    """
+
+    def __new__(cls, name, initial_size,
+                provisioned_size, content_type, format):
+        return tuple.__new__(cls, (
+            name, initial_size, provisioned_size, content_type, format))
+
+
+def register(parser):
+    cmd = parser.add_sub_command(
+        "upload-disk",
+        help="Upload disk",
+        func=upload_disk)
+
+    cmd.add_argument(
+        "-s", "--storage-domain",
+        required=True,
+        help="Name of the storage domain.")
+
+    cmd.add_argument(
+        "-f", "--format",
+        choices=("raw", "qcow2"),
+        default="qcow2",
+        help="Upload image format (default qcow2).")
+
+    cmd.add_argument(
+        "--sparse",
+        action="store_true",
+        help="Create sparse disk. Cannot be used with raw format on "
+             "block storage.")
+
+    cmd.add_argument(
+        "--id", "--disk-id",
+        type=_options.UUID,
+        help="A UUID for the new disk. If not specified oVirt will "
+             "create a new UUID.")
+
+    cmd.add_argument(
+        "--disable-backup",
+        dest="enable_backup",
+        action="store_false",
+        help="Creates a disk that can not be used for incremental backup.")
```

Comment:
Since this is not compatible with --format raw on any storage domain, we can validate it before creating a disk:

```
$ ./ovirt-img upload-disk -c engine -s fc-01 --format raw /data/tmp/fedora-35.qcow2
[ ---- ] 0 bytes, 0.13 s, 0 bytes/s | command failed                           
Traceback (most recent call last):
  File "/home/nsoffer/src/ovirt-imageio/./ovirt-img", line 13, in <module>
    _tool.main()
  File "/home/nsoffer/src/ovirt-imageio/ovirt_imageio/client/_tool.py", line 35, in main
    args.command(args)
  File "/home/nsoffer/src/ovirt-imageio/ovirt_imageio/client/_upload.py", line 83, in upload_disk
    disk = _ovirt.add_disk(
  File "/home/nsoffer/src/ovirt-imageio/ovirt_imageio/client/_ovirt.py", line 93, in add_disk
    disk = disks_service.add(
  File "/home/nsoffer/.local/lib/python3.10/site-packages/ovirtsdk4/services.py", line 7825, in add
    return self._internal_add(disk, headers, query, wait)
  File "/home/nsoffer/.local/lib/python3.10/site-packages/ovirtsdk4/service.py", line 232, in _internal_add
    return future.wait() if wait else future
  File "/home/nsoffer/.local/lib/python3.10/site-packages/ovirtsdk4/service.py", line 55, in wait
    return self._code(response)
  File "/home/nsoffer/.local/lib/python3.10/site-packages/ovirtsdk4/service.py", line 229, in callback
    self._check_fault(response)
  File "/home/nsoffer/.local/lib/python3.10/site-packages/ovirtsdk4/service.py", line 132, in _check_fault
    self._raise_error(response, body)
  File "/home/nsoffer/.local/lib/python3.10/site-packages/ovirtsdk4/service.py", line 118, in _raise_error
    raise error
ovirtsdk4.Error: Fault reason is "Operation Failed". Fault detail is "[Cannot add Virtual Disk. Disk configuration (RAW Preallocated backup-Incremental) is incompatible with the storage domain type.]". HTTP response code is 409.
(ovirt-imageio) [nsoffer@loop ovirt-imageio (aesteve/add-upload-disk-cmd)]$ ./ovirt-img upload-disk -c engine -s fc-01 --format raw 
```

We can add validate_args() helper at the start of the command, since argparse does not
support validation that depends on multiple arguments.

Reply:
We also have the case of uploading raw preallocated iamge to file based domain. This also fail since we enable_backup by default.

I think we need to modify enable backup based on the user request. If the user chose:

    --format raw

We cannot enable incremental backup, so we can change enable_backup to False.

In the UI the user cannot select the format - so we need the option to enable incremental backup.

These are the UI options:

- thin disk, block storage, enable backup -> sparse, qcow2
- preallocated disk, block storage, enable backup -> preallocated, qcow2
- thin disk, file storage, enable backup -> sparse, qcow2
- preallocated disk, file storage, enable backup -> preallocated qcow2
- thin disk, block storage -> sparse raw
- preallocated disk, block storage -> preallocated raw
- thin disk, file storage -> sparse, qcow2
- preallocated disk, file storage -> preallocated raw
 
The tool should make it easy to create all these combinations. The defaults should
create disks enabled for incremental backup. If the user select raw format we cannot
enable incremental backup.

We can require the user to use --format raw --disable-backup but this look too hard to use to me. We can warn that the disk cannot be used with incremental backup in this case.

Reply:
It would be nice if we can remove the --disable-backup flag, and use the enable_backup option automatically for the user.

The case that requires enable_backup is format=qcow2 sparse=False - engine does not allow this combination without enable_backup=True.